kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        rCloud_of_dofs[i] = nodes_array[i]->pGetDof(rVariable);
+        noalias(row(rCloud_of_nodes_coordinates,i)) = nodes_array[i]->Coordinates();
+    }
+    KRATOS_CATCH("");
+}
+
+
+void AssignMPCsToNeighboursUtility::GetDofsAndCoordinatesForNodes(
+    ResultNodesContainerType nodes_array,
+    const Variable<array_1d<double, 3>>& rVariable,
+    DofPointerVectorType& rCloud_of_dofs_x,
+    DofPointerVectorType& rCloud_of_dofs_y,
+    DofPointerVectorType& rCloud_of_dofs_z,
+    Matrix& rCloud_of_nodes_coordinates
+)
+{
+    KRATOS_TRY;
+    rCloud_of_dofs_x.resize(nodes_array.size());
+    rCloud_of_dofs_y.resize(nodes_array.size());
+    rCloud_of_dofs_z.resize(nodes_array.size());
+    rCloud_of_nodes_coordinates.resize(nodes_array.size(), 3);
+
+    for(int i = 0; i < static_cast<int>(nodes_array.size()); i++)
+    {
+        NodeType::Pointer pNode = nodes_array[i];
+
+        // Check if the node has the required DOFs
+        if (pNode->HasDofFor(GetComponentVariable(rVariable, 0)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 1)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
+
+            rCloud_of_dofs_x[i] = pNode->pGetDof(GetComponentVariable(rVariable, 0));
+            rCloud_of_dofs_y[i] = pNode->pGetDof(GetComponentVariable(rVariable, 1));
+            rCloud_of_dofs_z[i] = pNode->pGetDof(GetComponentVariable(rVariable, 2));
+
+        } else {
+            KRATOS_ERROR << "The node with ID " << pNode->Id() << " does not have the required DOFs for the variable " << rVariable << std::endl;
+        }
+        noalias(row(rCloud_of_nodes_coordinates, i)) = pNode->Coordinates();
+    }
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignRotationToNodes(
+    NodesContainerType pNodes,
+    Matrix RotationMatrix
+    )
+{
+    //TODO: Do it in parallel
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
+
+    // Assign rotation to given nodes
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        auto coordinate_vector = nodes_array[i]->GetInitialPosition().Coordinates();
+        Vector rotated_position(coordinate_vector.size());
+        noalias(rotated_position) = prod(RotationMatrix,coordinate_vector);
+        nodes_array[i]->X() = rotated_position[0];
+        nodes_array[i]->Y() = rotated_position[1];
+        nodes_array[i]->Z() = rotated_position[2];
+    }
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignMPCsToNodes(
+    NodesContainerType pNodes,
+    double const Radius,
+    ModelPart& rComputingModelPart,
+    const Variable<double>& rVariable,
+    double const MinNumOfNeighNodes
+    )
+{
+    KRATOS_TRY;
+
+    BuiltinTimer build_and_assign_mpcs;
+
+    int prev_num_mpcs = rComputingModelPart.NumberOfMasterSlaveConstraints();
```

Comment:
In here you are assuming that your new MPCs are not interacting with previous ones. Though this might be a valid assumption, I'd throw a warning message if the previous number of MPCs is non-zero to inform about this.